hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -298,14 +304,73 @@ func (r *NodePoolReconciler) reconcile(ctx context.Context, hcluster *hyperv1.Ho
 	}
 
 	r.Log.Info("Reconciling userdata Secret")
+	caSecret := ignitionserver.IgnitionCACertSecret(controlPlaneNamespace)
+	if err := r.Get(ctx, ctrlclient.ObjectKeyFromObject(caSecret), caSecret); err != nil {
+		if apierrors.IsNotFound(err) {
+			// TODO: Really want to just RequeueAfter, but the way the reconcile is
+			// structured eats the Result. Could also consider setting some condition
+			// so that a subsequent HostedCluster update would trigger a requeue.
+			return ctrl.Result{}, fmt.Errorf("ignition ca secret isn't yet available, waiting")
+		} else {
+			return ctrl.Result{}, fmt.Errorf("failed to get ca secret: %w", err)
+		}
+	}
+	tokenSecret := ignitionserver.IgnitionTokenSecret(controlPlaneNamespace)
+	if err := r.Get(ctx, ctrlclient.ObjectKeyFromObject(tokenSecret), tokenSecret); err != nil {
+		if apierrors.IsNotFound(err) {
+			// TODO: Really want to just RequeueAfter, but the way the reconcile is
```

Comment:
we can return result here https://github.com/openshift/hypershift/pull/275/files#diff-c666a2a9dc48b0d3de33cc8fa3a054f853671f0428f9d56b263750dfc6d26e00R194 if that's what we want. Though I'm not a fan of requeueAfter, it usually denotes a lack of watcher/design flaw, erroring seems fine to me here.

Reply:
You're right, this smells. Elsewhere in the hostedcluster controller I had a similar issue w/ the Route and added a hostedcluster parent annotation and a new watch for routes, and I can do the same thing here.

Reply:
Actually in this case it's not entirely clear what to watch to cause a non-error based retry... the hostedcluster-controller owns writing the IgnitionCACertSecret, and the nodepool-controller needs that secret to exist as input to certain recon tasks. What would the nodepool-controller watch to trigger a reconcile off IgnitionCACertSecret? The secret won't have the nodepool as a parent (via annotation), so watching secrets and enqueueing off parent in nodepool-controller doesn't seem right.

Returning an error when the secret is not found works but I'm not sure if it's a great solution.

What if the HostedCluster had an IgnitionServerAvailable condition? It could be set to True only when the server deployment is available (and thus all the secrets are implicitly available). Then the nodepool controller would resync when the HostedCluster status is updated with the new condition, and could even do a semantic gate on the condition before bothering to reconcile ignition stuff.

Reply:
I implemented status reporting for the ignition server based on its deployment, so now when that deployment is ready the HostedCluster's `IgnitionServerAvailable` condition should change and trigger the nodepool updates, allowing me to remove the error code and TODOs.